File: benchmarks/executable/restaurant_uint_wrapping_power.c
// C23 reference for the Restaurant UInt wrapping-power workload.

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile uint64_t runtime_base = UINT64_C(3);
static volatile uint64_t runtime_exponent = UINT64_C(40);

static uint64_t wrapping_power_u64(uint64_t base, uint64_t exponent) {
  uint64_t result = UINT64_C(1);
  while (exponent != UINT64_C(0)) {
    if ((exponent & UINT64_C(1)) != UINT64_C(0)) result *= base;
    base *= base;
    exponent >>= 1;
  }
  return result;
}

int main(void) {
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  const uint64_t wrapped =
      wrapping_power_u64(runtime_base, runtime_exponent);
  return printf("Wrapped %" PRIu64 "\n", wrapped) < 0 ? 1 : 0;
}

File: benchmarks/executable/restaurant_uint_wrapping_power.rs
// Rust 2024 reference for the Restaurant UInt wrapping-power workload.

use std::hint::black_box;
use std::io::{self, Write};

fn wrapping_power_u64(mut base: u64, mut exponent: u64) -> u64 {
    let mut result = 1_u64;
    while exponent != 0 {
        if exponent & 1 != 0 {
            result = result.wrapping_mul(base);
        }
        base = base.wrapping_mul(base);
        exponent >>= 1;
    }
    result
}

fn main() {
    let wrapped = wrapping_power_u64(black_box(3_u64), black_box(40_u64));
    let mut stdout = io::stdout().lock();
    if writeln!(stdout, "Wrapped {wrapped}").is_err() {
        std::process::exit(1);
    }
}
